### ISSUE TITLE
Speed up gradient norm clipping.

### ISSUE THREAD

**Issue by PhilJd** [PyTorchLightning/pytorch-lightning#2766]
## 🚀 Feature
Improve the speed of gradient clipping.

### Motivation
I've noticed that clipping the gradients is quite slow, e.g., for a ResNet50 model (with some minor extensions) gradient clipping takes ~207 ms. The major bottlenecks are the `torch.where` computation within a loop and the manual norm computation.

### Pitch
With slight modifications that don't hurt readability and work well on TPU,
grad_norm can be computed in ~75ms, i.e.,  2.7x as fast. For implementation details see the PR I'm going to link in a second.

Reproduce: Train any model you've got lying around with gradient_clip_val=0.1 and take a look at the trace timeline ;)

As a reference, here's a before and after comparison for my model (trace after 40 warmup steps):
Current:
![image](https://user-images.githubusercontent.com/16101605/88942052-09cd7100-d28a-11ea-93db-73c72e350ac4.png)

With speed improvements:
![image](https://user-images.githubusercontent.com/16101605/88941994-f4584700-d289-11ea-9c72-668a8b8d227e.png)
